File: src/main/python/LinkedListCycle.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast_pointer = head
        slow_pointer = head
        while fast_pointer is not None and fast_pointer.next is not None:
            fast_pointer = fast_pointer.next.next
            slow_pointer = slow_pointer.next
            if fast_pointer == slow_pointer:
                return True
        return False

File: src/main/java/solution/LinkedListCycle.java
package solution;

import datastructure.ListNode;

public class LinkedListCycle {
    // if pointer reaches null then it doesn't have cycle
    // if 2 pointers meet again, then it has cycle
    public static boolean sol(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
}
